CrashTracer: com.apple.WebKit.GPU at WebKit: WebKit::RemoteLegacyCDMSessionProxy::generateKeyRequest

youssefsoli opened the pull request:
#### dc4fdf81f20ec660ea5456b5b6b47d2357316f53
<pre>
CrashTracer: com.apple.WebKit.GPU at WebKit: WebKit::RemoteLegacyCDMSessionProxy::generateKeyRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=243554">https://bugs.webkit.org/show_bug.cgi?id=243554</a>
&lt;rdar://79306462&gt;

Reviewed by Jer Noble.

Throughout RemoteLegacyCDMSessionProxy, we assume that `cdm.createSession(*this)`
will not give us a null `m_session`, however, there exists implementations
of `createSession` that do return a nullptr. Since we never null check,
this results in a nullptr dereference in `RemoteLegacyCDMSessionProxy::generateKeyRequest`
and potentially everywhere `m_session` is dereferenced.

The fix is to null check everytime we want to use `m_session`.

* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
(WebKit::RemoteLegacyCDMSessionProxy::RemoteLegacyCDMSessionProxy):
(WebKit::RemoteLegacyCDMSessionProxy::generateKeyRequest):
(WebKit::RemoteLegacyCDMSessionProxy::releaseKeys):
(WebKit::RemoteLegacyCDMSessionProxy::update):
(WebKit::RemoteLegacyCDMSessionProxy::getCachedKeyForKeyId):
(WebKit::RemoteLegacyCDMSessionProxy::logChannel const):
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.h:
(WebKit::RemoteLegacyCDMSessionProxy::logClassName const):

Canonical link: <a href="https://commits.webkit.org/253152@main">https://commits.webkit.org/253152@main</a>
</pre>

youssefsoli commented on Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
```diff
@@ -46,15 +46,17 @@ std::unique_ptr<RemoteLegacyCDMSessionProxy> RemoteLegacyCDMSessionProxy::create
 }
 
-RemoteLegacyCDMSessionProxy::RemoteLegacyCDMSessionProxy(RemoteLegacyCDMFactoryProxy& factory, uint64_t logIdentifier, RemoteLegacyCDMSessionIdentifier sessionIdentifier, WebCore::LegacyCDM& cdm)
+RemoteLegacyCDMSessionProxy::RemoteLegacyCDMSessionProxy(RemoteLegacyCDMFactoryProxy& factory, uint64_t parentLogIdentifier, RemoteLegacyCDMSessionIdentifier sessionIdentifier, WebCore::LegacyCDM& cdm)
```
`logIdentifier` was renamed to avoid a naming conflict with the `logIndentifier()` call behind `LOGIDENTIFIER`

jernoble commented on Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
```diff
@@ -80,11 +83,12 @@ static RefPtr<WebCore::SharedBuffer> convertToOptionalSharedBuffer(T array)
 void RemoteLegacyCDMSessionProxy::generateKeyRequest(const String& mimeType, RefPtr<SharedBuffer>&& initData, GenerateKeyCallback&& completion)
 {
+    if (!m_session)
+        return completion({ }, emptyString(), 0, 0);
```
There was a long thread on webkit-dev about the appropriateness of `return void;`. I think the project as a whole decided against allowing this.

youssefsoli replied:
Sounds good

youssefsoli committed:
CrashTracer: com.apple.WebKit.GPU at WebKit: WebKit::RemoteLegacyCDMSessionProxy::generateKeyRequest
https://bugs.webkit.org/show_bug.cgi?id=243554
<rdar://79306462>

Reviewed by Jer Noble.

Throughout RemoteLegacyCDMSessionProxy, we assume that `cdm.createSession(*this)`
will not give us a null `m_session`, however, there exists implementations
of `createSession` that do return a nullptr. Since we never null check,
this results in a nullptr dereference in `RemoteLegacyCDMSessionProxy::generateKeyRequest`
and potentially everywhere `m_session` is dereferenced.

The fix is to null check everytime we want to use `m_session`.

* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.cpp:
(WebKit::RemoteLegacyCDMSessionProxy::RemoteLegacyCDMSessionProxy):
(WebKit::RemoteLegacyCDMSessionProxy::generateKeyRequest):
(WebKit::RemoteLegacyCDMSessionProxy::releaseKeys):
(WebKit::RemoteLegacyCDMSessionProxy::update):
(WebKit::RemoteLegacyCDMSessionProxy::getCachedKeyForKeyId):
(WebKit::RemoteLegacyCDMSessionProxy::logChannel const):
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMSessionProxy.h:
(WebKit::RemoteLegacyCDMSessionProxy::logClassName const):

Canonical link: https://commits.webkit.org/253152@main
